nixos/modules/config/system-environment.nix

```diff
@@ -23,7 +23,7 @@ in
         strings.  The latter is concatenated, interspersed with colon
         characters.
       '';
-      type = types.attrsOf (types.loeOf types.str);
+      type = with types; attrsOf (either str (listOf str));
```

Comment:
The merge problem was partly adressed in https://github.com/NixOS/nixpkgs/pull/20228.
It is not perfect as it will not merge variables if they have a different type, maybe that is the problem you are facing?